api/src/server/gql/mod.rs

```diff
@@ -104,35 +106,43 @@ impl QueryFields for Query {
 
 impl Cursor {
     fn from_index(index: i32) -> Self {
-        // TODO use base64
-        Self(format!("{}", index))
+        // i32 to base64 string
+        Self(base64::encode(index.to_be_bytes()))
     }
 
-    fn to_index(&self) -> ServerResult<i32> {
-        // TODO use base64
-        Ok(self.0.parse()?)
+    fn to_index(&self) -> Fallible<i32> {
+        // base64 string to i32. Convert to bytes first, then to int.
+        let decoded_bytes: Vec<u8> = base64::decode(&self.0)?;
+        let byte_array: [u8; 4] = decoded_bytes.as_slice().try_into()?;
+        Ok(i32::from_be_bytes(byte_array))
     }
 }
 
-impl Validate for Cursor {
-    fn validate(&self) -> Result<(), ValidationErrors> {
-        // we have to implement this manually because the struct definition
-        // is auto-generated, so we can't put more macros on it.
-        let mut error: ValidationError = match self.to_index() {
-            Ok(index) if index >= 0 => {
-                return Ok(());
-            }
-            _ => ValidationError::new("cursor"),
-        };
-        error.add_param("cursor".into(), &self.0);
-
-        // This error is kinda nasty but way she goes
-        let mut errors: ValidationErrors = ValidationErrors::new();
-        errors.add("__all__", error);
-        Err(errors)
+// We have to implement this manually because the struct definition is
+// auto-generated, so we can't put more macros on it. Needed for the validation
+// errors.
+impl Serialize for Cursor {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.0)
     }
 }
 
+fn validate_cursor(cursor: &Cursor) -> Result<(), ValidationError> {
+    // we have to implement this manually because the struct definition
+    // is auto-generated, so we can't put more macros on it.
+    let mut error: ValidationError = match cursor.to_index() {
+        Ok(index) if index >= 0 => {
+            return Ok(());
+        }
+        _ => ValidationError::new("cursor"),
+    };
+    error.add_param("message".into(), &"Invalid cursor value");
+    Err(error)
+}
+
```

Comment:
This fixes the nasty cursor validation error format from before. I had to manually implement `Serialize` for `Cursor` since the struct is generated by the GraphQL macro above, so I can't put `#[derive(Serialize)]` on it. Here's the improvement:

### Query

```graphql
{
  hardwareSpecs(after: "bad-cursor") {
    totalCount
  }
}
```

### Response Before

```json
{
  "data": null,
  "errors": [
    {
      "message": "Input validation error(s)",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "hardwareSpecs"
      ],
      "extensions": {
        "after": {
          "__all__": [
            {
              "cursor": "\"bad-cursor\""
            }
          ]
        }
      }
    }
  ]
}
```

### Response After

```json
{
  "data": null,
  "errors": [
    {
      "message": "Input validation error(s)",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "hardwareSpecs"
      ],
      "extensions": {
        "after": [
          {
            "message": "\"Invalid cursor value\"",
            "value": "\"bad-cursor\""
          }
        ]
      }
    }
  ]
}
```